File: modules/ducktests/tests/ignitetest/tests/cache_destroy.py
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
This module contains cache create/destroy tests that checks if
"""

from ducktape.mark.resource import cluster

from ignitetest.services.ignite_app import IgniteApplicationService
from ignitetest.services.ignite_execution_exception import IgniteExecutionException
from ignitetest.services.ignite import IgniteService
from ignitetest.services.ignite_app import IgniteApplicationService
from ignitetest.services.spark import SparkService
from ignitetest.services.utils.ignite_configuration.discovery import from_ignite_cluster
from ignitetest.services.utils.ignite_configuration import IgniteConfiguration
from ignitetest.services.zk.zookeeper import ZookeeperService
from ignitetest.utils import ignite_versions
from ignitetest.utils.ignite_test import IgniteTest
from ignitetest.utils.version import DEV_BRANCH, IgniteVersion

# pylint: disable=W0223
class CacheDestroyTest(IgniteTest):
    """
    https://sbtatlas.sigma.sbrf.ru/jira/browse/IGN-1794

    """
    NUM_NODES = 1
    CACHE_NAME = "TEST01"
    CACHES_AMOUNT = 10

    @cluster(num_nodes=NUM_NODES+1)
    @ignite_versions(str(DEV_BRANCH), str(LATEST_2_8))
    def test(self, ignite_version):

        node_config = IgniteConfiguration(version=IgniteVersion(ignite_version))

        ignites = IgniteService(self.test_context, config=node_config, num_nodes=self.NUM_NODES)
        ignites.start()

        # This client just put some data to the cache.
        app_config = node_config._replace(client_mode=True, discovery_spi=from_ignite_cluster(ignites))
        IgniteApplicationService(self.test_context, config=app_config,
                                 java_class_name="org.apache.ignite.internal.ducktest.tests.CreateDestroyCache",
                                 params={"cacheName": self.CACHE_NAME, "caches_number": self.CACHES_AMOUNT}).run()

File: modules/control-utility/src/main/java/org/apache/ignite/internal/commandline/cache/CacheDestroy.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.internal.commandline.cache;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;
import java.util.TreeSet;
import org.apache.ignite.IgniteLogger;
import org.apache.ignite.client.IgniteClient;
import org.apache.ignite.configuration.ClientConfiguration;
import org.apache.ignite.internal.commandline.AbstractCommand;
import org.apache.ignite.internal.commandline.Command;
import org.apache.ignite.internal.commandline.CommandArgIterator;
import org.apache.ignite.internal.commandline.TaskExecutor;
import org.apache.ignite.internal.util.typedef.F;
import org.apache.ignite.internal.util.typedef.internal.S;
import org.apache.ignite.internal.visor.cache.VisorCacheStopTask;
import org.apache.ignite.internal.visor.cache.VisorCacheStopTaskArg;

import static org.apache.ignite.internal.commandline.CommandLogger.or;
import static org.apache.ignite.internal.commandline.cache.CacheSubcommands.DESTROY;

/**
 * Cache destroy command.
 */
public class CacheDestroy extends AbstractCommand<VisorCacheStopTaskArg> {
    /** Command argument to destroy all user-created caches. */
    public static final String DESTROY_ALL_ARG = "--destroy-all-caches";

    /** Command argument to specify a comma-separated list of cache names to be destroyed. */
    public static final String CACHE_NAMES_ARG = "--caches";

    /** Confirmation message format. */
    public static final String CONFIRM_MSG = "Warning! The command will destroy %d caches: %s.\n" +
        "If you continue, the cache data will be impossible to recover.";

    /** No user-created caches exists message. */
    public static final String NOOP_MSG = "No user-created caches exist.";

    /** Result message. */
    public static final String RESULT_MSG = "The following caches have been stopped: %s.";

    /** Flag to destroy all user-created caches. */
    private boolean destroyAll;

    /** Caches to destroy. */
    private Set<String> cacheNames;

    /** {@inheritDoc} */
    @Override public VisorCacheStopTaskArg arg() {
        return new VisorCacheStopTaskArg(new ArrayList<>(cacheNames));
    }

    /** {@inheritDoc} */
    @Override public void printUsage(IgniteLogger log) {
        String cacheNamesArgFull = CACHE_NAMES_ARG + " cache1,...,cacheN";

        usageCache(
            log,
            DESTROY,
            "Permanently destroy specified caches.",
            F.asMap(cacheNamesArgFull, "specifies a comma-separated list of cache names to be destroyed.",
                DESTROY_ALL_ARG, "permanently destroy all user-created caches."),
            or(cacheNamesArgFull, DESTROY_ALL_ARG)
        );
    }

    /** {@inheritDoc} */
    @Override public void parseArguments(CommandArgIterator argIter) {
        cacheNames = null;
        destroyAll = false;

        String requiredArgsMsg = "One of \"" + CACHE_NAMES_ARG + "\" or \"" + DESTROY_ALL_ARG + "\" is expected.";

        String cmdArg = argIter.nextArg(requiredArgsMsg);

        if (DESTROY_ALL_ARG.equals(cmdArg))
            destroyAll = true;
        else if (CACHE_NAMES_ARG.equals(cmdArg))
            cacheNames = new TreeSet<>(argIter.nextStringSet("cache names"));
        else
            throw new IllegalArgumentException("Invalid argument \"" + cmdArg + "\". " + requiredArgsMsg);

        if (argIter.hasNextSubArg()) {
            throw new IllegalArgumentException(
                "Invalid argument \"" + argIter.peekNextArg() + "\", no more arguments is expected.");
        }
    }

    /** {@inheritDoc} */
    @Override public void prepareConfirmation(ClientConfiguration clientCfg) throws Exception {
        if (destroyAll)
            cacheNames = collectClusterCaches(clientCfg);
    }

    /** {@inheritDoc} */
    @Override public String confirmationPrompt() {
        if (F.isEmpty(cacheNames))
            return null;

        return String.format(CONFIRM_MSG, cacheNames.size(), S.joinToString(cacheNames, ", ", "..", 80, 0));
    }

    /** {@inheritDoc} */
    @Override public Object execute(ClientConfiguration clientCfg, IgniteLogger log) throws Exception {
        if (destroyAll && cacheNames == null)
            cacheNames = collectClusterCaches(clientCfg);

        if (cacheNames.isEmpty()) {
            assert destroyAll;

            log.info(NOOP_MSG);

            return null;
        }

        try (IgniteClient client = Command.startClient(clientCfg)) {
            TaskExecutor.executeTask(client, VisorCacheStopTask.class, arg(), clientCfg);
        }

        log.info(String.format(RESULT_MSG, F.concat(cacheNames, ", ")));

        return null;
    }

    /** {@inheritDoc} */
    @Override public String name() {
        return DESTROY.text().toUpperCase();
    }

    /**
     * @param clientCfg Client configuration.
     * @return Names of user-created caches that exist in the cluster.
     * @throws Exception If failed.
     */
    private Set<String> collectClusterCaches(ClientConfiguration clientCfg) throws Exception {
        try (IgniteClient client = Command.startClient(clientCfg)) {
            return new HashSet(client.cacheNames());
        }
    }
}
